### PR TITLE
Convert Pod ext resource to quota limit

### PR DESCRIPTION
**Pull request by chenkaiyue**
#### What this PR does / why we need it:
1. move node resource controller as node resource manager to agent
2. add low priority pod which used ext resource qos ensurance feature in node resource manager
3. add a data source, node resource manager can get data from both tsp and local

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -53,10 +56,12 @@ func NewAgent(ctx context.Context,
 
 	stateCollector := collector.NewStateCollector(nodeName, nepInformer.Lister(), podInformer.Lister(), nodeInformer.Lister(), ifaces, healthCheck, CollectInterval)
 	managers = append(managers, stateCollector)
-	analyzerManager := analyzer.NewAnormalyAnalyzer(kubeClient, nodeName, podInformer, nodeInformer, nepInformer, actionInformer, stateCollector.StateChann, noticeCh)
+	analyzerManager := analyzer.NewAnormalyAnalyzer(kubeClient, nodeName, podInformer, nodeInformer, nepInformer, actionInformer, stateCollector.AnalyzerChann, noticeCh)
 	managers = append(managers, analyzerManager)
 	avoidanceManager := executor.NewActionExecutor(kubeClient, nodeName, podInformer, nodeInformer, noticeCh, runtimeEndpoint)
 	managers = append(managers, avoidanceManager)
+	nodeResourceManager := node_resource.NewNodeResourceManager(kubeClient, nodeName, podInformer, nodeInformer, tspInformer, runtimeEndpoint, stateCollector.OfflineChann, stateCollector.Collectors)
```
Comment (yan234280533) [inline pkg/agent/agent.go:63]: Is here needs add an feature gate for the node resource ?

File: pkg/ensurance/collector/collector.go [hunk 2]
```diff
@@ -131,7 +134,8 @@ func (s *StateCollector) Collect() {
 
 	wg.Wait()
 
-	s.StateChann <- data
+	s.AnalyzerChann <- data
+	s.OfflineChann <- data
```
Comment (yan234280533) [inline pkg/ensurance/collector/collector.go:138]: rename it to NodeResouceChann ?

File: pkg/ensurance/collector/collector.go [hunk 3]
```diff
@@ -27,13 +27,15 @@ type StateCollector struct {
 	healthCheck     *metrics.HealthCheck
 	collectInterval time.Duration
 	ifaces          []string
-	collectors      *sync.Map
-	StateChann      chan map[string][]common.TimeSeries
+	Collectors      *sync.Map
```
Comment (yan234280533) [inline pkg/ensurance/collector/collector.go:30]: Why did we rename it to uppercase ？

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+	podSynced cache.InformerSynced
+
+	nodeLister corelisters.NodeLister
+	nodeSynced cache.InformerSynced
+
+	tspLister predictionlisters.TimeSeriesPredictionLister
+	tspSynced cache.InformerSynced
+
+	runtimeClient pb.RuntimeServiceClient
+	runtimeConn   *grpc.ClientConn
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:60]: It is not used , to remove it ?

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+	stateChann    chan map[string][]common.TimeSeries
+
+	// A copy of data from stateChann
+	state         map[string][]common.TimeSeries
+	// Updated when get new data from stateChann, used to determine whether state has expired
+	lastStateTime time.Time
+
+	collectors *sync.Map
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:68]: It is not used , to remove it ?

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+		collectors:    Collectors,
+	}
+	tspInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: o.reconcileTsp,
+		UpdateFunc: func(old, cur interface{}) {
+			o.reconcileTsp(cur)
+		},
+		DeleteFunc: o.reconcileTsp,
+	})
+	podInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:100]: need it moves to another manager ?

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+			return utils.GetContainerIdFromKey(cs.ContainerID)
+		}
+	}
+	return ""
+}
+
+// Get cpu period from local if cadvisor collector exists and state is not expired;
+// Otherwise, get value from CRI
+func (o *NodeResourceManager) getCPUPeriod(pod *v1.Pod, containerId string) float64 {
+	value, exists := o.collectors.Load(types.CadvisorCollectorType)
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:345]: Can it judge from the state?

Comment (yan234280533) [reply pkg/ensurance/node-resource/node-resource-manager.go:345]: Not to use collect direct.

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+					continue
+				}
+
+				containerPeriod := o.getCPUPeriod(pod, containerId)
+				if containerPeriod == 0 {
+					continue
+				}
+
+				// Update cpu quota by CRI
+				err := cruntime.UpdateContainerResources(o.runtimeClient, containerId, cruntime.UpdateOptions{CPUQuota: int64(float64(val.MilliValue()) / executor.CpuQuotaCoefficient * containerPeriod)})
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:323]: If the throttle modified the pod, as it is conflict ?

File: pkg/ensurance/node-resource/node-resource-manager.go [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+			if period.ContainerId == containerId {
+				return period.Value
+			}
+		}
+	}
+
+	// Use CRI to get cpu period directly
+	c := value.(cadvisor.CadvisorCollector)
+	var query = info.ContainerInfoRequest{}
+	containerInfoV1, err := c.Manager.GetContainerInfo(containerId, &query)
```
Comment (yan234280533) [inline pkg/ensurance/node-resource/node-resource-manager.go:360]: Not get state direct, we can use the state

File: pkg/agent/agent.go (hunk 1 above)
Comment (chenkaiyue) [reply pkg/agent/agent.go:63]: ok

File: pkg/ensurance/collector/collector.go (hunk 3 above)
Comment (chenkaiyue) [reply pkg/ensurance/collector/collector.go:30]: node resource manager need to import it

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 5 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go:68]: We use it in getCPUPeriod

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 7 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go:345]: The purpose here is to get period from state first , if state is not ok ,then we should directly get it by CRI.

File: pkg/ensurance/collector/collector.go (hunk 2 above)
Comment (chenkaiyue) [reply pkg/ensurance/collector/collector.go:138]: ok

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 4 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go:60]: It is used in updatePodExtResToCgroup.

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 6 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go:100]: ok

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 9 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go:360]: We inded get cpu period from local if cadvisor collector exists and state is not expired first.
If state is not available, get it from CRI.

File: pkg/ensurance/executor/throttle.go [hunk 10]
```diff
@@ -281,7 +281,7 @@ func (t *ThrottleExecutor) Restore(ctx *ExecuteContext) error {
 						bSucceed = false
 						continue
 					}
-					klog.V(2).Infof("restore kkkkkkkk")
+					klog.V(2).Infof("restore")
```
Comment (mfanjie) [inline pkg/ensurance/executor/throttle.go:284]: log should be a full statement.
and v(2) can be removed per code standard.

File: pkg/resource/node-resource-manager.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+	predictionv1 "github.com/gocrane/api/pkg/generated/informers/externalversions/prediction/v1alpha1"
+	predictionlisters "github.com/gocrane/api/pkg/generated/listers/prediction/v1alpha1"
+	predictionapi "github.com/gocrane/api/prediction/v1alpha1"
+	"github.com/gocrane/crane/pkg/common"
+	"github.com/gocrane/crane/pkg/prediction/config"
+)
+
+const (
+	ExtResourcePrefixFormat = "ext-resource.node.gocrane.io/%s"
```
Comment (mfanjie) [inline pkg/resource/node-resource-manager.go:30]: is `gocrane.io/%s` better name?

File: pkg/resource/pod-resource-manger.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+				if containerId == "" {
+					continue
+				}
+
+				containerPeriod := o.getCPUPeriod(pod, containerId)
+				if containerPeriod == 0 {
+					continue
+				}
+
+				// Update cpu quota by CRI
```
Comment (mfanjie) [inline pkg/resource/pod-resource-manger.go:160]: should you compare quota before update?
we should avoid cgroup change on every pod update event.

File: pkg/resource/pod-resource-manger.go [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+			}
+		}
+	}
+
+	// Use CRI to get cpu period directly
+	c := value.(cadvisor.CadvisorCollector)
+	var query = info.ContainerInfoRequest{}
+	containerInfoV1, err := c.Manager.GetContainerInfo(containerId, &query)
+	if err != nil {
+		klog.Errorf("ContainerInfoRequest failed: %v", err)
```
Comment (mfanjie) [inline pkg/resource/pod-resource-manger.go:200]: for which containerid?

File: pkg/resource/pod-resource-manger.go [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	go func() {
+		for {
+			select {
+			case state := <-o.stateChann:
+				o.state = state
+				o.lastStateTime = time.Now()
+			case <-stop:
+				klog.Infof("pod resource manager exit")
```
Comment (mfanjie) [inline pkg/resource/pod-resource-manger.go:111]: log starts with Capital letter.

**Comment by mfanjie** [comment]
@szy441687879 can you confirm this change, node resource controller is moving to agent by this pr, please confirm there is no duplicate work here.

**Comment by mfanjie** [comment]
@shijieqin moved the node resource controller in https://github.com/gocrane/crane/pull/198 as well.

File: pkg/resource/node-resource-manager.go [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+			case v1.ResourceCPU:
+				// cpu need to be scaled to m as ext resource cannot be decimal
+				nextRecommendation = (float64(node.Status.Allocatable.Cpu().Value()) - maxUsage) * 1000
+			case v1.ResourceMemory:
+				// unit of memory in prometheus is in Ki, need to be converted to byte
+				nextRecommendation = float64(node.Status.Allocatable.Memory().Value()) - (maxUsage * 1000)
+			default:
+				continue
+			}
+			if nextRecommendation <= 0 {
```
Comment (yan234280533) [inline pkg/resource/node-resource-manager.go:216]: If the recommendation is zero, set the extend resource to zero is more better ?

File: pkg/resource/pod-resource-manger.go [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/gocrane/crane/pkg/ensurance/collector/cadvisor"
+	"github.com/gocrane/crane/pkg/ensurance/collector/types"
+	stypes "github.com/gocrane/crane/pkg/ensurance/collector/types"
+	"github.com/gocrane/crane/pkg/ensurance/executor"
+	cgrpc "github.com/gocrane/crane/pkg/ensurance/grpc"
+	cruntime "github.com/gocrane/crane/pkg/ensurance/runtime"
+	"github.com/gocrane/crane/pkg/utils"
+)
+
+type PodResourceManager struct {
```
Comment (yan234280533) [inline pkg/resource/pod-resource-manger.go:30]: pod-resource-manger.go => pod_resource_manger.go , the golang file name standard ？

File: pkg/resource/node-resource-manager.go [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+				klog.Errorf("Failed to update node %s's status extend-resource, %v", nodeCopy.Name, err)
+				return
+			}
+			klog.V(4).Infof("Update Node %s Extend Resource Success according to TSP %s", node.Name, tsp.Name)
+		}
+		return
+	}
+}
+
+func (o *NodeResourceManager) getNode() (*v1.Node, []v1.NodeAddress) {
```
Comment (yan234280533) [inline pkg/resource/node-resource-manager.go:150]: Not return NodeAddress here, we can get address from the node struct

File: pkg/resource/node-resource-manager.go [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+		}
+	}()
+
+	return
+}
+
+func (o *NodeResourceManager) UpdateNodeResource() {
+	tsps, err := o.tspLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("List tsp error: %#v", err)
```
Comment (yan234280533) [inline pkg/resource/node-resource-manager.go:108]: klog.Errorf("List tsp error: %#v", err) => klog.Errorf("Failed to list tsp: %v", err)

File: pkg/resource/node-resource-manager.go [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+	stateChann chan map[string][]common.TimeSeries
+
+	//TODO: use state to be a backup of tsp
+	// A copy of data from stateChann
+	state map[string][]common.TimeSeries
+	// Updated when get new data from stateChann, used to determine whether state has expired
+	lastStateTime time.Time
+}
+
+func NewNodeResourceManager(client clientset.Interface, nodeName string, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer,
```
Comment (yan234280533) [inline pkg/resource/node-resource-manager.go:55]: we neead add metrics for the node resource manager

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 8 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go]: So shall we chose a smaller one between limit and the current quota?

File: pkg/resource/node-resource-manager.go (hunk 11 above)
Comment (chenkaiyue) [reply pkg/resource/node-resource-manager.go:30]: Ok

File: pkg/ensurance/executor/throttle.go (hunk 10 above)
Comment (chenkaiyue) [reply pkg/ensurance/executor/throttle.go:284]: Ok

File: pkg/resource/pod-resource-manger.go (hunk 13 above)
Comment (chenkaiyue) [reply pkg/resource/pod-resource-manger.go:200]: ok

File: pkg/resource/pod-resource-manger.go (hunk 14 above)
Comment (chenkaiyue) [reply pkg/resource/pod-resource-manger.go:111]: ok

File: pkg/resource/node-resource-manager.go (hunk 15 above)
Comment (chenkaiyue) [reply pkg/resource/node-resource-manager.go:216]: OK, i correct it to <

File: pkg/resource/pod-resource-manger.go (hunk 16 above)
Comment (chenkaiyue) [reply pkg/resource/pod-resource-manger.go:30]: Ok

File: pkg/resource/node-resource-manager.go (hunk 17 above)
Comment (chenkaiyue) [reply pkg/resource/node-resource-manager.go:150]: Ok

File: pkg/resource/node-resource-manager.go (hunk 18 above)
Comment (chenkaiyue) [reply pkg/resource/node-resource-manager.go:108]: ok

File: pkg/ensurance/node-resource/node-resource-manager.go (hunk 8 above)
Comment (chenkaiyue) [reply pkg/ensurance/node-resource/node-resource-manager.go]: 1. -1 -> quota
2. 限制最大改到quota

File: pkg/resource/pod-resource-manger.go (hunk 12 above)
Comment (chenkaiyue) [reply pkg/resource/pod-resource-manger.go:160]: I discussed with devin, pod resource manager only convert containers whose quota is -1, if it has value, the adjustment can handled by throttle module. And throttle module has the logic about compare.

File: pkg/resource/node-resource-manager.go (hunk 19 above)
Comment (chenkaiyue) [reply pkg/resource/node-resource-manager.go]: ok

File: pkg/ensurance/executor/throttle.go [hunk 20]
```diff
@@ -256,9 +258,17 @@ func (t *ThrottleExecutor) Restore(ctx *ExecuteContext) error {
 					containerCPUQuotaNew = float64(limitCPU.MilliValue()) / CpuQuotaCoefficient
 				}
 			} else {
-				if containerCPUQuotaNew > MaxUpQuota*containerCPUPeriod.Value/CpuQuotaCoefficient {
+				hasExtRes := false
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/executor/throttle.go:261]: ok

File: pkg/ensurance/collector/collector.go [hunk 21]
```diff
@@ -124,7 +130,9 @@ func (s *StateCollector) Collect() {
 
 	wg.Wait()
 
-	s.StateChann <- data
+	s.AnalyzerChann <- data
+	s.NodeResourceChann <- data
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/collector/collector.go:134]: i add a judgement here.

File: pkg/resource/pod_resource_manger.go [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+				if err != nil {
+					klog.Errorf("Failed to update pod %s container %s Resource, err %s", pod.Name, containerId, err.Error())
+					continue
+				}
+			}
+		}
+	}
+}
+
+func getContainerIdFromPod(pod *v1.Pod, containerName string) string {
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/resource/pod_resource_manger.go:179]: ok

File: pkg/resource/pod_resource_manger.go [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+
+				containerPeriod := o.getCPUPeriod(pod, containerId)
+				if containerPeriod == 0 {
+					continue
+				}
+
+				// Update cpu quota by CRI
+				err = cruntime.UpdateContainerResources(o.runtimeClient, containerId, cruntime.UpdateOptions{CPUQuota: int64(float64(val.MilliValue()) / executor.CpuQuotaCoefficient * containerPeriod)})
+				if err != nil {
+					klog.Errorf("Failed to update pod %s container %s Resource, err %s", pod.Name, containerId, err.Error())
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/resource/pod_resource_manger.go:176]: ok

File: pkg/agent/agent.go [hunk 24]
```diff
@@ -52,11 +57,21 @@ func NewAgent(ctx context.Context,
 
 	stateCollector := collector.NewStateCollector(nodeName, nepInformer.Lister(), podInformer.Lister(), nodeInformer.Lister(), ifaces, healthCheck, CollectInterval)
 	managers = append(managers, stateCollector)
-	analyzerManager := analyzer.NewAnormalyAnalyzer(kubeClient, nodeName, podInformer, nodeInformer, nepInformer, actionInformer, stateCollector.StateChann, noticeCh)
+	analyzerManager := analyzer.NewAnormalyAnalyzer(kubeClient, nodeName, podInformer, nodeInformer, nepInformer, actionInformer, stateCollector.AnalyzerChann, noticeCh)
 	managers = append(managers, analyzerManager)
 	avoidanceManager := executor.NewActionExecutor(kubeClient, nodeName, podInformer, nodeInformer, noticeCh, runtimeEndpoint)
 	managers = append(managers, avoidanceManager)
 
+	if nodeResource := utilfeature.DefaultFeatureGate.Enabled(features.CraneNodeResource); nodeResource {
```
Comment (qmhu) [inline pkg/agent/agent.go:65]: Do you init featuregate before use it ?
`utilfeature.DefaultMutableFeatureGate.AddFlag(cmd.Flags())`

File: pkg/resource/node_resource_manager.go [hunk 25]
```diff
@@ -0,0 +1,232 @@
...
+				continue
+			}
+			extResourceName := fmt.Sprintf(executor.ExtResourcePrefixFormat, string(*resourceName))
+			resValue, exists := node.Status.Capacity[v1.ResourceName(extResourceName)]
+			if exists && resValue.Value() != 0 &&
+				math.Abs(float64(resValue.Value())-
+					nextRecommendation)/float64(resValue.Value()) <= MinDeltaRatio {
+				continue
+			}
+			node.Status.Capacity[v1.ResourceName(extResourceName)] =
```
Comment (qmhu) [inline pkg/resource/node_resource_manager.go:226]: if resource is cpu ，should we use binarySI here？

File: pkg/utils/ref.go [hunk 26]
```diff
@@ -36,3 +36,12 @@ func GetContainerIdFromKey(key string) string {
 
 	return ""
 }
+
+func GetContainerIdFromPod(pod *v1.Pod, containerName string) string {
+	for _, cs := range pod.Status.ContainerStatuses {
```
Comment (qmhu) [inline pkg/utils/ref.go:41]: better to move it to pkg/utils/pod.go

File: pkg/agent/agent.go (hunk 24 above)
Comment (chenkaiyue) [reply pkg/agent/agent.go:65]: I put it in defaultFeatureGates and use true as default

Comment (chenkaiyue) [reply pkg/agent/agent.go:65]: Ok, i add it into agent command.

File: pkg/resource/node_resource_manager.go (hunk 25 above)
Comment (chenkaiyue) [reply pkg/resource/node_resource_manager.go:226]: cpu should use DecimalSI in my opinion?

File: pkg/utils/ref.go (hunk 26 above)
Comment (chenkaiyue) [reply pkg/utils/ref.go:41]: ok

File: pkg/resource/node_resource_manager.go (hunk 25 above)
Comment (qmhu) [reply pkg/resource/node_resource_manager.go:231]: ok, what if the resource is memory?

Comment (chenkaiyue) [reply pkg/resource/node_resource_manager.go:231]: BinarySI

File: pkg/resource/node-resource-manager.go (hunk 15 above)
Comment (shijieqin) [reply pkg/resource/node-resource-manager.go]: If the recommendation is less than zero, the extension resource should also be set to zero, right? In the presence of a watermark, the recommendation is likely to be less than 0

File: pkg/resource/pod_resource_manger.go [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+			}
+		}
+	}
+	metrics.UpdateDurationFromStart(string(known.ModulePodResourceManager), metrics.StepUpdatePodResource, start)
+}
+
+// Get cpu period from local if cadvisor collector exists and state is not expired;
+// Otherwise, get value from CRI
+func (o *PodResourceManager) getCPUPeriod(pod *v1.Pod, containerId string) float64 {
+	value, exists := o.collectors.Load(types.CadvisorCollectorType)
```
Comment (szy441687879) [inline pkg/resource/pod_resource_manger.go:194]: collectors may not have CadvisorCollectorType?

File: pkg/resource/pod_resource_manger.go [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+		_, containerCPUPeriods := executor.GetPodUsage(string(stypes.MetricNameContainerCpuPeriod), o.state, pod)
+		for _, period := range containerCPUPeriods {
+			if period.ContainerId == containerId {
+				return period.Value
+			}
+		}
+	}
+
+	// Use CRI to get cpu period directly
+	c := value.(cadvisor.CadvisorCollector)
```
Comment (szy441687879) [inline pkg/resource/pod_resource_manger.go:207]: value.(*cadvisor.CadvisorCollector)？

**Comment by chenkaiyue** [comment]
image
image
image

**Pull request merged**